Fix spurious "File Changed" dialogs using `hash` from `jupyter-server` v2.11.1+

Wh1isper opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

Fixes https://github.com/jupyterlab/jupyterlab/issues/7743
Fixes https://github.com/jupyterlab/jupyterlab/issues/11352

References PR: https://github.com/jupyterlab/jupyterlab/pull/15037

Upstream:
- https://github.com/jupyter-server/jupyter_server/pull/1363
- https://github.com/jupyter-server/jupyter_server/pull/1367

## Code changes

<!-- Describe the code changes and how they address the issue. -->

Since jupyter server may provide hash in model, we compare hash for the file changes first.

**Forgive me for being a TypeScript and front-end novice, more advice would be appreciated, thanks!**

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

N/A

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

N/A

krassowski commented on packages/docregistry/src/context.ts:
```diff
@@ -780,13 +827,13 @@ or load the version on disk (revert)?`,
       actions: ['overwrite']
     });
-    this._timeConflictModalIsOpen = true;
+    this._ConflictModalIsOpen = true;
```
```suggestion
    this._conflictModalIsOpen = true;
```

krassowski commented on packages/docregistry/src/context.ts:
```diff
@@ -780,13 +827,13 @@ or load the version on disk (revert)?`,
       buttons: [Dialog.cancelButton(), revertBtn, overwriteBtn]
     }).then(result => {
-      this._timeConflictModalIsOpen = false;
+      this._ConflictModalIsOpen = false;
```
```suggestion
      this._conflictModalIsOpen = false;
```

krassowski commented on packages/docregistry/src/context.ts:
```diff
@@ -917,7 +964,7 @@ or load the version on disk (revert)?`,
   private _dialogs: ISessionContext.IDialogs;
   private _lastModifiedCheckMargin = 500;
-  private _timeConflictModalIsOpen = false;
+  private _ConflictModalIsOpen = false;
```
```suggestion
  private _conflictModalIsOpen = false;
```

krassowski commented on packages/docregistry/src/context.ts:
```diff
@@ -746,22 +801,14 @@ export class Context<
-        `${tDisk}`
-    );
-    if (this._timeConflictModalIsOpen) {
+    if (this._ConflictModalIsOpen) {
```
```suggestion
    if (this._conflictModalIsOpen) {
```

krassowski reviewed (changes requested):
Thank you for this PR! It looks good except for one condition which I am not sure about.

Would you consider also adding a test for this? I think it could be added along:

https://github.com/jupyterlab/jupyterlab/blob/6b498b46e98dd796e6b3022363907ba97de5bce8/packages/docregistry/test/context.spec.ts#L85-L108

Wh1isper commented on packages/docregistry/src/context.ts:
```diff
@@ -476,11 +476,19 @@ export class Context<
+    if (
+      !mod ||
+      newModel.last_modified !== mod ||
+      !hash ||
+      newModel.hash !== hash
```
Yes, we should assume that it is possible that `hash` does not exist(is `null`). This is not the same as `mod`.

Wh1isper commented:
I will add a test on this soon, thanks!

Wh1isper commented on packages/docregistry/test/context.spec.ts:
I'm not sure such a test is complete or enough. Feel free to give more feedback!

Wh1isper commented on packages/docregistry/test/context.spec.ts:
I'm not sure such a test is complete or enough. Feel free to give more feedback!

krassowski commented on packages/docregistry/test/context.spec.ts:
```diff
@@ -85,10 +85,15 @@ describe('docregistry/context', () => {
     describe('#fileChanged', () => {
       it('should be emitted when the file is saved', async () => {
         const path = context.path;
+        const hash = context.contentsModel?.hash ?? null;
         let called = false;
         context.fileChanged.connect((sender, args) => {
           expect(sender).toBe(context);
           expect(args.path).toBe(path);
+          if (hash) {
+            expect(args.hash).not.toBeNull();
+            expect(args.hash).not.toBe(hash);
+          }
```
I think that rather than using a conditional expect a new dedicated test would be desirable. In particular such a test should verify that the `context.fileChanged` is not emitted. It will need some work to mock the server response here I believe.

Wh1isper commented:
I changed some logic about emit `fileChanged` and added `hash` and `hash_algorithm` in mock.

I have tested following situations manually: 
- when `hash` is provided, `fileChanged` is not emitted.
- when `hash` is not provided, `fileChanged` is emitted because of `timeStamp` changed.

But I don't know how to reorganise the test cases to support both scenarios.

Feel free to give me more feed back, thanks!

krassowski reviewed (approved):
Thank you @Wh1isper!

I pushed a few more tests and tiny adjustments in the conditionals.
